[wasm] SIMD support improvements

radekdoulik opened the pull request:
Add internal PackedSimd class with wasm specific SIMD intrinsics methods.

radekdoulik committed:
Initial wasm SIMD support

radekdoulik committed:
Enable Vector intrinsic on wasm

The llvm code generator works nicely with them.

radekdoulik committed:
Add missing files

radekdoulik committed:
Make SIMD support conditional

radekdoulik committed:
Remove test code

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-2

radekdoulik committed:
Fix debug build

radekdoulik committed:
Update after merge

radekdoulik committed:
Add Splat and ExcractLane methods

radekdoulik committed:
Switch i64 values for Constant method

So that C#

    WasmBase.Constant(0xff11ff22ff33ff44, 0xff55ff66ff77ff88)

is compiled into wasm code

    v128.const 0xff11ff22ff33ff44ff55ff66ff77ff88    [SIMD]

radekdoulik committed:
Update PlatformNotSupported version of WasmBase

radekdoulik committed:
Fix CI build

radekdoulik committed:
Add ReplaceLane and Swizzle

radekdoulik committed:
Change WasmBase.Constant to get Vector128 as input

radekdoulik committed:
Add Shuffle methods

This will need more work, as it crashes clang during 'WebAssembly
Instruction Selection' pass:

    WasmApp.Native.targets(353,5): error : 3.    Running pass 'WebAssembly Instruction Selection' on function '@corlib_System_Runtime_Intrinsics_Wasm_WasmBase_Shuffle_System_Runtime_Intrinsics_Vector128_1_byte_System_Runtime_Intrinsics_Vector128_1_byte_System_Runtime_Intrinsics_Vector128_1_byte'

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-2

radekdoulik committed:
Handle SN_Shuffle

radekdoulik committed:
Fix crash in OP_STOREX_MEMBASE

radekdoulik committed:
Add build test

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-2

radekdoulik committed:
Fix remaining conflict

radekdoulik committed:
Remove unused prop

radekdoulik committed:
IsSupported should be static

radekdoulik committed:
Handle SN_get_IsSupported

radekdoulik committed:
Keep passing --enable-simd to wasm-opt

radekdoulik committed:
Add Span reverse measurements

radekdoulik committed:
Use WasmBase to improve perf of Span reverse

| measurement | no SIMD | SIMD |
|-:|-:|-:|
|                    Span, Reverse bytes |     0.0341ms |     0.0028ms |
|                    Span, Reverse chars |     0.0394ms |     0.0062ms |

radekdoulik committed:
Remove debug prints

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-4

radekdoulik committed:
Do not expose WasmBase API

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-4

radekdoulik committed:
Add BuildAdditionalArgs prop

To pass args to sample builds

radekdoulik committed:
Handle WasmBase.get_IsSupported in the interpreter

radekdoulik committed:
Add IndexOf measurements

radekdoulik committed:
Add And intrinsic methods

radekdoulik committed:
Add more methods to WasmBase classes

radekdoulik committed:
Add implementation of the new WasmBase methods

radekdoulik committed:
Use WasmBase to improve Span.IndexOf performance

Measurements chrome/aot/amd64:

| measurement | no SIMD | SIMD |
|-:|-:|-:|
|                    Span, Reverse bytes |     0.0332ms |     0.0025ms |
|                    Span, Reverse chars |     0.0332ms |     0.0060ms |
|                    Span, IndexOf bytes |     0.2068us |     0.1002us |
|                    Span, IndexOf chars |     0.0146ms |     0.0028ms |

Measurements firefox/aot/amd64:

| measurement | no SIMD | SIMD |
|-:|-:|-:|
|                    Span, Reverse bytes |     0.0338ms |     0.0022ms |
|                    Span, Reverse chars |     0.0339ms |     0.0048ms |
|                    Span, IndexOf bytes |     0.2533us |     0.1394us |
|                    Span, IndexOf chars |     0.0201ms |     0.0039ms |

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-4

radekdoulik commented:
/azp run runtime-wasm

vargaz commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -2443,6 +2443,7 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 	} else if (in_corlib &&
 			(!strncmp ("System.Runtime.Intrinsics.Arm", klass_name_space, 29) ||
+			!strncmp ("System.Runtime.Intrinsics.WasmBase", klass_name_space, 29) ||
 			!strncmp ("System.Runtime.Intrinsics.X86", klass_name_space, 29)) &&
```
The length might not be correct.

vargaz commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -9683,7 +9683,49 @@ MONO_RESTORE_WARNING
 #endif
-
+#if defined(TARGET_WASM)
+		case OP_WASM_SIMD_BITMASK: {
```
This could use the OP_XOP_X_X opcode, there is not need for a separate one.

radekdoulik replied:
I didn't find any suitable XOP for bitmask and swizzle intrinsics. The llvm has intrinsic functions for these too. Maybe I am looking at wrong places? Which X opcodes would fit for these?

vargaz replied:
Looks like it doesn't exist, but swizzle is defined as:
def int_wasm_swizzle :
  Intrinsic<[llvm_v16i8_ty],
            [llvm_v16i8_ty, llvm_v16i8_ty],
            [IntrNoMem, IntrSpeculatable]>;

So it's just a binary operation on vectors, so it could use OP_XOP_X_X_X.

radekdoulik replied:
I didn't find any suitable XOP. The wasm swizzle instruction (2 vectors shuffle, 2nd value acts as mask) is similar to sse shuffle. We are using sse special OPs for sse shuffle too (OP_SSE_SHUFPS, OP_SSE2_PSHUFD, OP_SSE2_SHUFPD). So I guess we don't have OP_XOP_... for shuffle.

IIRC, I also tried to use `LLVMBuildShuffleVector` at some point, which didn't yield the swizzle instruction selection for me, so I ended calling the intrinsic function. The llvm codegen tests also call it. The tests also test use of series of extractelement/insertelement, but that is not what we want.

vargaz commented on src/mono/mono/mini/mini-llvm.c:
```diff
@@ -9683,7 +9683,49 @@ MONO_RESTORE_WARNING
+			break;
+		}
+		case OP_WASM_SIMD_SWIZZLE: {
+			LLVMValueRef args [] = { lhs, rhs };
```
Ditto, this could use OP_XOP_X_X_X.

vargaz commented on src/mono/mono/mini/simd-intrinsics.c:
```diff
@@ -4001,6 +4001,75 @@ emit_amd64_intrinsics (const char *class_ns, const char *class_name, MonoCompile
+		case SN_CompareNotEqual:
+			return emit_simd_ins_for_sig (cfg, klass, type_enum_is_float (arg0_type) ? OP_XCOMPARE_FP : OP_XCOMPARE, CMP_NE, arg0_type, fsig, args);
+		case SN_Constant:
+			return emit_simd_ins_for_sig (cfg, klass, OP_WASM_SIMD_V128_CONST, 0, arg0_type, fsig, args);
```
Why is this needed ?

radekdoulik replied:
Ah, that is not really needed, the OP_WASM_SIMD_V128_CONST remained here from older iterations. I will try to get rid of it.

radekdoulik replied:
I have removed OP_WASM_SIMD_V128_CONST

vargaz commented:
The runtime changes look ok otherwise.

radekdoulik committed:
Feedback, fix string length in the comparison

tannergooding commented:
Any perf numbers for these changes?

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:
```diff
@@ -587,7 +588,7 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
                 }
             }
-            else if (Sse2.IsSupported)
+            else if (Sse2.IsSupported || WasmBase.IsSupported)
```
If the `WASM` hookups also existed for the `Vector128<T>` APIs, then this could be `Vector128.IsHardwareAccelerated`

and then the line below could just be `int matches = Vector128.CompareEqual(values, search).ExtractMostSignificantBits();`

The code would then work for x86, x64, Arm64, and WASM without needing to have many separate paths.

radekdoulik replied:
The changes here to SpanHelper's are to demonstrate the `WasmBase` functionality and improve Span performance on wasm. They are quite small and done in similar way as other platforms here, so I hope they don't make the situation worse.

It will be nice to make these helpers use Vector128<T> for all the platforms you mentioned. Lets do that in separate PR? I can help with it.

For this PR I aim to get partial WasmBase implementation and speedup the Span on wasm. I am keeping the WasmBase internal for now, so that I can make changes if something doesn't go as desired. I plan to get it to API review once it is complete. I didn't propose any API for review yet, so hope this is good plan.

tannergooding replied:
> It will be nice to make these helpers use Vector128<T> for all the platforms you mentioned. Lets do that in separate PR?

Sounds good. Like I mentioned below, this is what we're hoping to do for more paths overall since it's less code to support more platforms and with the same perf for most scenarios.

>  am keeping the WasmBase internal for now, so that I can make changes if something doesn't go as desired. I plan to get it to API review once it is complete.

Yep, that's definitely good since we can't expose any new public API without it going through review first.

> I didn't propose any API for review yet, so hope this is good plan.

https://github.com/dotnet/runtime/issues/53730 should be the existing API proposal, so we shouldn't need anything new. Just update this one with any tweaks necessary before we take it to review.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs:
```diff
@@ -2374,7 +2381,7 @@ public static void Reverse(ref byte buf, nuint length)
                 length -= numIters * numElements * 2;
             }
-            else if (Ssse3.IsSupported && (nuint)Vector128<byte>.Count * 2 <= length)
+            else if ((Ssse3.IsSupported || WasmBase.IsSupported) && (nuint)Vector128<byte>.Count * 2 <= length)
```
Same here, but using `Vector128.Shuffle(tempFirst, Vector128.Create((byte)15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0));` and `Vector128.Shuffle(tempLast, Vector128.Create((byte)15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0));`

The `Create` calls do need to be "inline" due to a JIT limitation in .NET 7, but they will get optimized and hoisted regardless since we have the relevant `GT_CNS_VEC` support in RyuJIT (Mono LLVM has something similar).

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Wasm/WasmBase.cs:
```diff
@@ -0,0 +1,198 @@
+namespace System.Runtime.Intrinsics.Wasm
+{
+    [Intrinsic]
+    internal abstract class WasmBase
```
We should take this (well https://github.com/dotnet/runtime/issues/53730) to actual API review for .NET 8.

Provided that the relevant hookups exist for the xplat `Vector128<T>` (and `Vector<T>`) APIs, WASM will also be able to light up on any paths using those APIs prior to WASM support being directly exposed.

radekdoulik replied:
> Provided that the relevant hookups exist for the xplat `Vector128<T>` (and `Vector<T>`) APIs, WASM will also be able to light up on any paths using those APIs prior to WASM support being directly exposed.

IIRC we already handle xplat for Vector128<T> on wasm. Vector<T> is next for me to look into, I hope I can speedup string and other parts on wasm.

tannergooding replied:
Thanks for the info!

We're looking at moving more code to use `Vector128<T>` xplat APIs where possible so that we can support more platforms with less code overall (and only using the platform specific APIs where it gives significant improvement or where functionality can't be represented using the xplat APIs).

So supporting `Vector128<T>` will allow more WASM light-up faster.

radekdoulik committed:
Feedback, remove OP_WASM_SIMD_V128_CONST

radekdoulik commented:
> Any more extensive perf numbers for these changes?

I can add more wasm measurents, if that is what you are looking for?

> We have many paths which use the `Vector128<T>` (and several which still use `Vector<T>`) xplat APIs now and which could get WASM lightup provided the recognition existed.

We already have some of `Vector128<T>` intrinsics in place for wasm.

tannergooding commented:
> I can add more wasm measurents, if that is what you are looking for?

I misunderstood exactly what lightup this was bringing at first. I thought it was adding `Vector128<T>` light up as well and so was expecting perf numbers for more than just `Span<T>`.

Now that I understand its just adding `WasmBase` and using that directly for `Span<T>` where we were already using xplat intrinsics, I think the numbers provided are fine.

radical commented:
@radekdoulik btw, you can use `/azp run runtime-wasm-perf` to run all the wasm benchmarks from `dotnet/performance` with this PR.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Wasm/WasmBase.PlatformNotSupported.cs:
```diff
@@ -0,0 +1,97 @@
+
+namespace System.Runtime.Intrinsics.Wasm
+{
+    internal abstract class WasmBase
```
Not critical for this PR, but just noting that the approved API surface differs slightly from what the proposal initially had: https://github.com/dotnet/runtime/issues/53730#issuecomment-1209770214

`WasmBase -> PackedSimd`, `Constant` is cut as you should just use `Vector128.Create(...)` instead, most other things just had the parameter names updated from `x`, `a`, `b`, and `imm` to be things like `left`, `right`, `value`, `index`, etc.

radekdoulik replied:
> @radekdoulik btw, you can use `/azp run runtime-wasm-perf` to run all the wasm benchmarks from `dotnet/performance` with this PR.

nice to have this pipeline. it would not measure these changes though, as they are hidden behind `WasmSIMD` property, which is disabled by default.

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-4

radekdoulik committed:
Merge remote-tracking branch 'remotes/origin/main' into pr-wasm-simd-wip-4

radekdoulik committed:
s/WasmBase/PackedSimd

And update parameter names and indentation

radekdoulik committed:
Remove Constant method implementation

radekdoulik commented:
Nice surprise to have the API approved! :-) I have updated the PR to reflect the changes and removed the span helper changes, as @adamsitnik's change uses the Vector128 now. The perf is not yet where it was, I will address that with next PR.

I am moving the old measurements here from PR description to keep them for later comparison.

Improve Span Reverse and IndexOf performance on wasm with SIMD

Measurements chrome/aot/amd64:

| measurement | no SIMD | SIMD |
|-:|-:|-:|
|                    Span, Reverse bytes |     0.0332ms |     0.0025ms |
|                    Span, Reverse chars |     0.0332ms |     0.0060ms |
|                    Span, IndexOf bytes |     0.2068us |     0.1002us |
|                    Span, IndexOf chars |     0.0146ms |     0.0028ms |

Measurements firefox/aot/amd64:

| measurement | no SIMD | SIMD |
|-:|-:|-:|
|                    Span, Reverse bytes |     0.0338ms |     0.0022ms |
|                    Span, Reverse chars |     0.0339ms |     0.0048ms |
|                    Span, IndexOf bytes |     0.2533us |     0.1394us |
|                    Span, IndexOf chars |     0.0201ms |     0.0039ms |

Code size reduction, the browser-bench sample: -5184 bytes (no SIMD 19539839 --> SIMD 19534655)

radekdoulik committed:
Remove `using` clauses

radekdoulik committed:
One more place to remove Constant method

radekdoulik committed:
Fix white space

radekdoulik commented:
I think I addressed all things from reviews. Is there anything else to change? I would like it to merge before it becomes outdated again ;-)

radical commented:
/azp run runtime-wasm,runtime-wasm-perf

radical commented:
Running `runtime-wasm-perf` just as a sanity check for the perf pipeline.
